File: src/part01/chapter02/levelC/matrix/java/MatrixCloner.java
package part01.chapter02.levelC.matrix.java;

public class MatrixCloner {

    public static double[][] cloneMatrix(double[][] matrixForClone) {
        double[][] result = new double[matrixForClone.length][];
        for (int rowIndex = 0; rowIndex < result.length; rowIndex++) {
            result[rowIndex] = matrixForClone[rowIndex].clone();
        }
        return result;
    }
}

File: src/part01/chapter02/levelC/matrix/kotlin/MatrixCloner.kt
package part01.chapter02.levelC.matrix.kotlin

fun cloneMatrix(matrix: Array<DoubleArray>): Array<DoubleArray> {
    val result = Array(matrix.size) { doubleArrayOf()}
    for (rowIndex in matrix.indices) {
        result[rowIndex] = matrix[rowIndex].clone()
    }
    return result
}